src/test/java/unit/uk/gov/ida/verifyserviceprovider/services/UnsignedAssertionResponseHandlerTest.java

```diff
@@ -0,0 +1,233 @@
+                )
+        );
+        Assertion eidasSamlAssertion = anEidasSamlAssertion(singleKeyList);
+
+        assertThrows(SamlFailedToDecryptException.class, () -> {
+            handler.decryptAssertion(validatedResponse, eidasSamlAssertion);
+        });
+    }
+
+    private Assertion anEidasSamlAssertion(List<String> keys) {
+        return anAssertion()
+                .addAttributeStatement(
+                        anAttributeStatement()
+                                .addAttribute(createCountrySamlResponseAttribute(samlString))
+                                .addAttribute(createEncryptedAssertionKeysAttribute(keys))
+                                .build())
+                .buildUnencrypted();
+    }
+
+    private Attribute createCountrySamlResponseAttribute(String countrySaml) {
+        CountrySamlResponse attributeValue = new CountrySamlResponseBuilder().buildObject();
+        attributeValue.setValue(countrySaml);
+
+        Attribute attribute = (Attribute) XMLObjectSupport.buildXMLObject(Attribute.DEFAULT_ELEMENT_NAME);
+        attribute.setName(IdaConstants.Eidas_Attributes.UnsignedAssertions.EidasSamlResponse.NAME);
+        attribute.setFriendlyName(IdaConstants.Eidas_Attributes.UnsignedAssertions.EidasSamlResponse.FRIENDLY_NAME);
+        attribute.setNameFormat(Attribute.URI_REFERENCE);
+
+        attribute.getAttributeValues().add(attributeValue);
+        return attribute;
+    }
+
+    private Attribute createEncryptedAssertionKeysAttribute(List<String> keys) {
+        List<EncryptedAssertionKeys> assertionKeysValues = new ArrayList<>();
+        for (String key : keys) {
+            EncryptedAssertionKeys attributeValue = new EncryptedAssertionKeysBuilder().buildObject();
+            attributeValue.setValue(key);
+            assertionKeysValues.add(attributeValue);
+        }
+
+        Attribute attribute = (Attribute) XMLObjectSupport.buildXMLObject(Attribute.DEFAULT_ELEMENT_NAME);
+        attribute.setName(IdaConstants.Eidas_Attributes.UnsignedAssertions.EncryptedSecretKeys.NAME);
+        attribute.setFriendlyName(IdaConstants.Eidas_Attributes.UnsignedAssertions.EncryptedSecretKeys.FRIENDLY_NAME);
+        attribute.setNameFormat(Attribute.URI_REFERENCE);
+
+        attribute.getAttributeValues().addAll(assertionKeysValues);
+        return attribute;
+    }
+
+    private Response createEidasResponse() throws Exception {
+        return aResponse()
+            .addEncryptedAssertion(
+                    anEidasAssertion().build())
+            .withInResponseTo(inResponseTo)
+            .build();
+    }
+
+    private EncryptionAlgorithmValidator getEncryptionAlgorithmValidator() {
+        return new EncryptionAlgorithmValidator(
+                ImmutableSet.of(
+                        EncryptionConstants.ALGO_ID_BLOCKCIPHER_AES128
+                ),
+                ImmutableSet.of(
+                        EncryptionConstants.ALGO_ID_KEYTRANSPORT_RSAOAEP
+                )
+        );
```

Comment:
Rather than an `EidasEncryptionAlgorithmValidator`? I guess this is more restrictive - so perhaps a better test?

Reply:
The crypto setup that the tests use, use this (`ALGO_ID_BLOCKCIPHER_AES128`) encryption algorithm. This is different to the algorithms specified in `EidasEncryptionAlgorithmValidator` which is why the tests use the validator being returned here.

We test that the handler will validate assertions encrypted with the algorithm it's configured with, not specifically that its the algorithms that the eIDAS spec specifies.

There's another test, `decryptAssertionShouldThrowIfWrongEncryptionAlgorithmUsed` that ensures that the handler throws if it encounters an assertion encrypted with an algorithm different to those listed in its validator.